File: examples/cpp/mgpcg.cpp
#include <taichi/lang.h>
#include <taichi/util.h>
#include <taichi/visual/gui.h>
#include <taichi/system/profiler.h>
#include <taichi/visual/texture.h>

TC_NAMESPACE_BEGIN

using namespace Tlang;

constexpr int dim = 3,
              n = 512;  // double the size for padding. actual box is 256^3
constexpr int pre_and_post_smoothing = 2, bottom_smoothing = 50;
constexpr int mg_levels = 5;

auto mgpcg_poisson = [](std::vector<std::string> cli_param) {
  auto param = parse_param(cli_param);

  int block_size = 8;

  int threads = param.get("threads", 8);
  TC_P(threads);
  bool vec_option = param.get("vec", true);
  int vec = vec_option ? block_size : 1;
  TC_ASSERT(vec == 1 || vec == block_size);
  TC_P(vec);
  bool load_gt = param.get("load_gt", false);
  TC_P(load_gt)
  bool gpu = param.get("gpu", false);
  TC_P(gpu)

  CoreState::set_trigger_gdb_when_crash(true);

  Program prog(gpu ? Arch::gpu : Arch::x86_64);
  // prog.config.print_ir = true;

  prog.config.simplify_before_lower_access = param.get("simp1", true);
  TC_P(prog.config.simplify_before_lower_access);
  prog.config.lower_access = param.get("lower_access", true);
  TC_P(prog.config.lower_access);
  prog.config.print_ir = param.get("print_ir", false);
  TC_P(prog.config.print_ir);
  prog.config.simplify_after_lower_access = param.get("simp2", true);
  TC_P(prog.config.simplify_after_lower_access);
  prog.config.attempt_vectorized_load_cpu = param.get("vec_load_cpu", true);
  TC_P(prog.config.attempt_vectorized_load_cpu);

  prog.config.lazy_compilation = false;

  auto r = Vector(DataType::f32, mg_levels);
  auto z = Vector(DataType::f32, mg_levels);
  Global(x, f32);
  Global(p, f32);
  Global(Ap, f32);
  Global(alpha, f32);
  Global(beta, f32);
  Global(sum, f32);
  Global(phase, i32);

  bool block_soa = true;

  layout([&]() {
    auto ijk = Indices(0, 1, 2);
    std::function<void(Expr & expr)> place_scalar;

    SNode *block;
    if (block_soa) {
      block = &root.dense(ijk, n / block_size).bitmasked();  //.pointer();
      place_scalar = [&](Expr &s) { block->dense(ijk, block_size).place(s); };
    } else {
      place_scalar = [&](Expr &mat) {
        root.dense(ijk, n / block_size).dense(ijk, block_size).place(mat);
      };
    }

    place_scalar(x);
    place_scalar(p);
    place_scalar(Ap);
    place_scalar(r(0));
    place_scalar(z(0));
    for (int i = 1; i < mg_levels; i++) {
      auto &fork = root.dense(ijk, n / block_size).bitmasked();
      fork.dense(ijk, block_size).place(r(i));
      fork.dense(ijk, block_size).place(z(i));
    }
    root.place(alpha, beta, sum, phase);
  });

  Kernel(compute_Ap).def([&] {
    // BlockDim(1024);
    Parallelize(threads);
    Vectorize(block_size);
    For(Ap, [&](Expr i, Expr j, Expr k) {
      Ap[i, j, k] = 6.0f * p[i, j, k] - p[i - 1, j, k] - p[i + 1, j, k] -
                    p[i, j + 1, k] - p[i, j - 1, k] - p[i, j, k + 1] -
                    p[i, j, k - 1];
    });
  });

  Kernel(reduce_r).def([&] {
    mark_reduction();
    Parallelize(threads);
    Vectorize(block_size);
    For(p, [&](Expr i, Expr j, Expr k) {
      Atomic(sum[Expr(0)]) += r(0)[i, j, k] * r(0)[i, j, k];
    });
  });

  Kernel(reduce_zTr).def([&] {
    mark_reduction();
    Parallelize(threads);
    Vectorize(block_size);
    For(p, [&](Expr i, Expr j, Expr k) {
      Atomic(sum[Expr(0)]) += z(0)[i, j, k] * r(0)[i, j, k];
    });
  });

  Kernel(reduce_pAp).def([&] {
    mark_reduction();
    Parallelize(threads);
    Vectorize(block_size);
    For(p, [&](Expr i, Expr j, Expr k) {
      auto tmp = Var(0.0f);
      tmp += p[i, j, k] * Ap[i, j, k];
      Atomic(sum[Expr(0)]) += tmp;
    });
  });

  Kernel(update_x).def([&] {
    Parallelize(threads);
    Vectorize(block_size);
    For(x, [&](Expr i, Expr j, Expr k) {
      x[i, j, k] += alpha[Expr(0)] * p[i, j, k];
    });
  });

  Kernel(update_r).def([&] {
    Parallelize(8);
    Vectorize(block_size);
    For(p, [&](Expr i, Expr j, Expr k) {
      r(0)[i, j, k] -= alpha[Expr(0)] * Ap[i, j, k];
    });
  });

  Kernel(update_p).def([&] {
    Parallelize(threads);
    Vectorize(block_size);
    For(p, [&](Expr i, Expr j, Expr k) {
      p[i, j, k] = z(0)[i, j, k] + beta[Expr(0)] * p[i, j, k];
    });
  });

  int begin = n / 4, end = n * 3 / 4;
  for (int i = begin; i < end; i++) {
    for (int j = begin; j < end; j++) {
      for (int k = begin; k < end; k++) {
        /*
        r(0).val<float32>(i, j, k) =
            (i == n / 2) && (j == n / 2) && (k == n / 2);
            */
        float x = (i - begin) * 2.0f / n;
        float y = (j - begin) * 2.0f / n;
        float z = (k - begin) * 2.0f / n;
        r(0).val<float32>(i, j, k) =
            sin(2 * pi * x) * cos(2 * pi * y) * sin(2 * pi * z);
      }
    }
  }

  std::vector<std::function<void()>> smoothers(mg_levels),
      smoothers2(mg_levels), restrictors(mg_levels - 1),
      prolongators(mg_levels - 1), clearer_z(mg_levels), clearer_r(mg_levels);

  for (int l = 0; l < mg_levels; l++) {
    if (l < mg_levels - 1) {
      restrictors[l] =
          kernel([&] {
            kernel_name(fmt::format("restrict_lv{}", l));
            Parallelize(threads);
            Vectorize(1);
            For(r(l), [&](Expr i, Expr j, Expr k) {
              auto res =
                  Var(r(l)[i, j, k] - (6.0f * z(l)[i, j, k] -
                                       z(l)[i - 1, j, k] - z(l)[i + 1, j, k] -
                                       z(l)[i, j + 1, k] - z(l)[i, j - 1, k] -
                                       z(l)[i, j, k + 1] - z(l)[i, j, k - 1]));
              if (gpu)
                Atomic(r(l + 1)[i / 2, j / 2, k / 2]) += res * 0.5f;
              else
                r(l + 1)[i / 2, j / 2, k / 2] += res * 0.5f;
            });
          })
              .func();
      prolongators[l] = kernel([&] {
                          kernel_name(fmt::format("prolongate_lv{}", l));
                          Parallelize(8);
                          Vectorize(1);
                          For(z(l), [&](Expr i, Expr j, Expr k) {
                            z(l)[i, j, k] += z(l + 1)[i / 2, j / 2, k / 2];
                          });
                        })
                            .func();
    }
    // prog.config.print_ir = true;
    smoothers[l] =
        kernel([&] {
          kernel_name(fmt::format("smooth_lv{}", l));
          Parallelize(threads);
          Vectorize(block_size);
          For(z(l), [&](Expr i, Expr j, Expr k) {
            auto ret = Var(0.0f);
            If(((i + j + k) && 1) == phase[Expr(0)]).Then([&] {
              ret = (r(l)[i, j, k] + z(l)[i - 1, j, k] + z(l)[i + 1, j, k] +
                     z(l)[i, j + 1, k] + z(l)[i, j - 1, k] + z(l)[i, j, k + 1] +
                     z(l)[i, j, k - 1]) *
                        (1.0f / 6) -
                    z(l)[i, j, k];
            });
            z(l)[i, j, k] += ret;
          });
        })
            .func();
    prog.config.print_ir = false;
    clearer_r[l] =
        kernel([&] {
          kernel_name(fmt::format("clear_r_lv{}", l));
          Parallelize(threads);
          Vectorize(block_size);
          For(r(l), [&](Expr i, Expr j, Expr k) { r(l)[i, j, k] = 0.0f; });
        })
            .func();
    clearer_z[l] =
        kernel([&] {
          kernel_name(fmt::format("clear_z_lv{}", l));
          Parallelize(threads);
          Vectorize(block_size);
          For(z(l), [&](Expr i, Expr j, Expr k) { z(l)[i, j, k] = 0.0f; });
        })
            .func();
  }
  Kernel(identity).def([&] {
    Parallelize(threads);
    Vectorize(block_size);
    For(z(0), [&](Expr i, Expr j, Expr k) { z(0)[i, j, k] = r(0)[i, j, k]; });
  });

  // z = M^-1 r
  auto apply_preconditioner = [&] {
    clearer_z[0]();
    for (int l = 0; l < mg_levels - 1; l++) {
      for (int i = 0; i < (pre_and_post_smoothing << l); i++) {
        phase.val<int32>() = 0;
        smoothers[l]();
        phase.val<int32>() = 1;
        smoothers[l]();
      }
      clearer_z[l + 1]();
      clearer_r[l + 1]();
      restrictors[l]();
    }
    for (int i = 0; i < bottom_smoothing; i++) {
      phase.val<int32>() = 0;
      smoothers[mg_levels - 1]();
      phase.val<int32>() = 1;
      smoothers[mg_levels - 1]();
    }
    for (int l = mg_levels - 2; l >= 0; l--) {
      prolongators[l]();
      for (int i = 0; i < (pre_and_post_smoothing << l); i++) {
        phase.val<int32>() = 1;
        smoothers[l]();
        phase.val<int32>() = 0;
        smoothers[l]();
      }
    }
  };

  sum.val<float32>() = 0;
  reduce_r();
  auto initial_rTr = sum.val<float32>();
  TC_P(initial_rTr);

  // r = b - Ax = b    since x = 0
  // p = r = r + 0 p
  apply_preconditioner();
  update_p();
  sum.val<float32>() = 0;
  reduce_zTr();
  auto old_zTr = sum.val<float32>();

  // CG
  auto t = Time::get_time();
  for (int i = 0; i < 400; i++) {
    TC_P(i);
    compute_Ap();
    sum.val<float32>() = 0;
    reduce_pAp();
    auto pAp = sum.val<float32>();
    // alpha = rTr / pTAp
    alpha.val<float32>() = old_zTr / pAp;
    // TC_P(old_zTr);
    // TC_P(pAp);
    // TC_P(alpha.val<float32>());
    // x = x + alpha p
    update_x();
    // r = r - alpha Ap
    update_r();
    // return if |r| small
    // z = M^-1 r
    apply_preconditioner();
    sum.val<float32>() = 0;
    reduce_zTr();
    auto new_zTr = sum.val<float32>();
    // TC_P(new_zTr);
    sum.val<float32>() = 0;
    reduce_r();
    auto rTr = sum.val<float32>();
    TC_P(rTr);
    if (rTr < initial_rTr * 1e-12f)
      break;
    // beta = new rTr / old rTr
    beta.val<float32>() = new_zTr / old_zTr;
    // TC_P(beta.val<float32>());
    // p = z + beta p
    update_p();
    old_zTr = new_zTr;
  }
  TC_P(Time::get_time() - t);
  get_current_program().profiler_print();

  compute_Ap();
  auto residual = 0.0f;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      for (int k = 0; k < n; k++) {
        for (int d = 0; d < dim; d++) {
          // residual += std::abs(R[i][j][k][d] - Ap(d).val<float32>(i, j, k));
        }
      }
    }
  }
  TC_P(residual);
  // TC_P(difference_max);

  std::vector<float32> ref_input(pow<3>(n / 2));
  if (load_gt) {
    auto f = fopen("mgpcg_ref.bin", "rb");
    fread(ref_input.data(), sizeof(float), ref_input.size(), f);
  }
  float absmax = 0;
  for (auto r : ref_input) {
    absmax = std::max(std::abs(absmax), r);
  }
  TC_P(absmax);

  int gui_res = 512;
  GUI gui("MGPCG Poisson", Vector2i(gui_res + 200, gui_res), false);
  int gt = 0;
  int k = 0;
  gui.slider("z", k, n / 4, n * 3 / 4 - 1).slider("Ground truth", gt, 0, 1);

  int scale = gui_res / n * 2;
  auto &canvas = gui.get_canvas();
  for (int frame = 1;; frame++) {
    for (int i = 0; i < gui_res - scale; i++) {
      for (int j = 0; j < gui_res - scale; j++) {
        real dx;
        if (!gt) {
          dx = x.val<float32>(i / scale + n / 4, j / scale + n / 4, k) / 256;
        } else {
          dx = ref_input[((i / scale) * n / 2 + j / scale) * n / 2 + k] / 256;
        }
        canvas.img[i][j] = Vector4(0.5f) + Vector4(dx, dx, dx, 0);
      }
    }
    gui.update();
  }
};
TC_REGISTER_TASK(mgpcg_poisson);

TC_NAMESPACE_END

File: examples/mgpcg.py
import numpy as np
import taichi as ti

real = ti.f32
ti.init(default_fp=real, arch=ti.x64, enable_profiler=True)

# grid parameters
N = 128
N_gui = 512  # gui resolution

n_mg_levels = 4
pre_and_post_smoothing = 2
bottom_smoothing = 50

use_multigrid = True

N_ext = N // 2  # number of ext cells set so that that total grid size is still power of 2
N_tot = 2 * N

# setup sparse simulation data arrays
r = [ti.var(dt=real) for _ in range(n_mg_levels)]  # residual
z = [ti.var(dt=real) for _ in range(n_mg_levels)]  # M^-1 r
x = ti.var(dt=real)  # solution
p = ti.var(dt=real)  # conjugate gradient
Ap = ti.var(dt=real)  # matrix-vector product
alpha = ti.var(dt=real)  # step size
beta = ti.var(dt=real)  # step size
sum = ti.var(dt=real)  # storage for reductions
pixels = ti.var(dt=real, shape=(N_gui, N_gui))  # image buffer

grid = ti.root.pointer(ti.ijk, [N_tot // 4]).dense(ti.ijk, 4).place(x, p, Ap)

for l in range(n_mg_levels):
    grid = ti.root.pointer(ti.ijk,
                           [N_tot // (4 * 2**l)]).dense(ti.ijk,
                                                        4).place(r[l], z[l])

ti.root.place(alpha, beta, sum)


@ti.kernel
def init():
    for i, j, k in ti.ndrange((N_ext, N_tot - N_ext), (N_ext, N_tot - N_ext),
                              (N_ext, N_tot - N_ext)):
        xl = (i - N_ext) * 2.0 / N_tot
        yl = (j - N_ext) * 2.0 / N_tot
        zl = (k - N_ext) * 2.0 / N_tot
        r[0][i, j, k] = ti.sin(2.0 * np.pi * xl) * ti.sin(
            2.0 * np.pi * yl) * ti.sin(2.0 * np.pi * zl)
        z[0][i, j, k] = 0.0
        Ap[i, j, k] = 0.0
        p[i, j, k] = 0.0
        x[i, j, k] = 0.0


@ti.kernel
def compute_Ap():
    for i, j, k in Ap:
        Ap[i,j,k] = 6.0 * p[i,j,k] - p[i+1,j,k] - p[i-1,j,k] \
                                   - p[i,j+1,k] - p[i,j-1,k] \
                                   - p[i,j,k+1] - p[i,j,k-1]


@ti.kernel
def reduce(p: ti.template(), q: ti.template()):
    for I in ti.grouped(p):
        sum[None] += p[I] * q[I]


@ti.kernel
def update_x():
    for I in ti.grouped(p):
        x[I] += alpha[None] * p[I]


@ti.kernel
def update_r():
    for I in ti.grouped(p):
        r[0][I] -= alpha[None] * Ap[I]


@ti.kernel
def update_p():
    for I in ti.grouped(p):
        p[I] = z[0][I] + beta[None] * p[I]


@ti.kernel
def restrict(l: ti.template()):
    for i, j, k in r[l]:
        res = r[l][i, j, k] - (6.0 * z[l][i, j, k] - z[l][i + 1, j, k] -
                               z[l][i - 1, j, k] - z[l][i, j + 1, k] -
                               z[l][i, j - 1, k] - z[l][i, j, k + 1] -
                               z[l][i, j, k - 1])
        r[l + 1][i // 2, j // 2, k // 2] += res * 0.5


@ti.kernel
def prolongate(l: ti.template()):
    for I in ti.grouped(z[l]):
        z[l][I] = z[l + 1][I // 2]


@ti.kernel
def smooth(l: ti.template(), phase: ti.template()):
    # phase = red/black Gauss-Seidel phase
    for i, j, k in r[l]:
        if (i + j + k) & 1 == phase:
            z[l][i,j,k] = (r[l][i,j,k] + z[l][i+1,j,k] + z[l][i-1,j,k] \
                                       + z[l][i,j+1,k] + z[l][i,j-1,k] \
                                       + z[l][i,j,k+1] + z[l][i,j,k-1])/6.0


def apply_preconditioner():
    z[0].fill(0)
    for l in range(n_mg_levels - 1):
        for i in range(pre_and_post_smoothing << l):
            smooth(l, 0)
            smooth(l, 1)
        z[l + 1].fill(0)
        r[l + 1].fill(0)
        restrict(l)

    for i in range(bottom_smoothing):
        smooth(n_mg_levels - 1, 0)
        smooth(n_mg_levels - 1, 1)

    for l in reversed(range(n_mg_levels - 1)):
        prolongate(l)
        for i in range(pre_and_post_smoothing << l):
            smooth(l, 1)
            smooth(l, 0)


@ti.kernel
def paint():
    kk = N_tot * 3 // 8
    for i, j in pixels:
        ii = int(i * N / N_gui) + N_ext
        jj = int(j * N / N_gui) + N_ext
        pixels[i, j] = x[ii, jj, kk] / N_tot


gui = ti.GUI("mgpcg", res=(N_gui, N_gui))

init()

sum[None] = 0.0
reduce(r[0], r[0])
initial_rTr = sum[None]

# r = b - Ax = b    since x = 0
# p = r = r + 0 p
if use_multigrid:
    apply_preconditioner()
else:
    z[0].copy_from(r[0])

update_p()

sum[None] = 0.0
reduce(z[0], r[0])
old_zTr = sum[None]

# CG
for i in range(400):
    # alpha = rTr / pTAp
    compute_Ap()
    sum[None] = 0.0
    reduce(p, Ap)
    pAp = sum[None]
    alpha[None] = old_zTr / pAp

    # x = x + alpha p
    update_x()

    # r = r - alpha Ap
    update_r()

    # check for convergence
    sum[None] = 0.0
    reduce(r[0], r[0])
    rTr = sum[None]
    if rTr < initial_rTr * 1.0e-12:
        break

    # z = M^-1 r
    if use_multigrid:
        apply_preconditioner()
    else:
        z[0].copy_from(r[0])

    # beta = new_rTr / old_rTr
    sum[None] = 0.0
    reduce(z[0], r[0])
    new_zTr = sum[None]
    beta[None] = new_zTr / old_zTr

    # p = z + beta p
    update_p()
    old_zTr = new_zTr

    print(' ')
    print(i)
    print(rTr)
    paint()
    gui.set_image(pixels)
    gui.show()

ti.profiler_print()
